components/WalletProvider.tsx

```diff
@@ -16,60 +16,52 @@ type WalletProviderProps = {
   children?: React.ReactNode
 }
 
+let provider: ethers.providers.Web3Provider
+let chainId: number
+let signer: Signer | undefined
```

Comment:
Moving these variables outside of the React state may cause the app to not react to changes in chain or signer. That feels dangerous.

Reply:
We have an action handler on chain change and account change, which directly leads to disconnect and going through the flow again. I don't see them adding any value in the state and causing unnecessary re-renders for the whole app, yes if in the future we support a chain change without disconnect it might make sense to have it in the state, but for now, I feel this is what we should keep. Let me know your thoughts

Reply:
I've tested disconnecting and changing chains, and everything does work as expected.

If we believe that taking these variables out of state is going to lead to a meaningful performance improvement, we should just call it out with a big bold comment around both the variables and the disconnect action that currently triggers the re-render. That way a future developer won't accidentally refactor the code and wind up with an application that is out of sync with the wallet provider.

If it's not a significant performance improvement, I'd suggest we keep everything in state for the sake of maintainability and simplicity.

Reply:
Umm saves us at least a few re-renders, umm around 5-7 in all. I will for now add a comment there, like you suggested so in future if someone wishes to refactor they do consider it and then do changes accordingly